Locking issue when reading data

There seems to be some kind of locking issue in Pathfinder, when I start a transaction (for example mint), while that mint tx is in pending state on mainnet, I can't read totalSupply() for that same token (tx fails), it seems like mint tx locks totalSupply storage value, and I can't read it while tx that is modifying that value is pending. I'm not sure if this is caused because Infura is using v0.4.0 pathfinder that is updated to cairo v0.10.2 and mainnet is still on v0.10.1, or something else. I tested same thing with sequencer and everything is working properly.

So, with the help of @lazarmitic we've been able to dig further into the issue.

This is the mint transaction: https://starkscan.co/tx/0x3c00fd81ca2416e783bc47cff4863ed6c7e904d51886fca90121f733646bce7

We've observed the following:

- Pathfinder periodically updates the pending block from the gateway. As soon as the transaction above made it into pathfinder's pending block the `starknet_call` to `totalSupply` of contract 0x070f8a4fcd75190661ca09a7300b7c93fab93971b67ea712c664d7948a8a54c6 started failing.
- Once the pending block has made it onto the chain (so the transaction above was L2_ACCEPTED) the same `starknet_call` call started working again.

The exact same issue can be reproduced by the following command (and a sufficiently up-to-date mainnet database containing block 11486):

```python
    c = Call(
        at_block="0x74cc50f4b8083835682bbe1489a4b1474053e4f6a35c66b8bb25b2e700630f9",
        chain=call.Chain.MAINNET,
        pending_updates={
            0x3D39F7248FB2BFB960275746470F7FB470317350AD8656249EC66067559E892: [
                call.StorageDiff(
                    key=0x57E3AC4C831E9CE78DE8142D212350256E7263262456048615D1800F91A021,
                    value=0xDE21C10E8E2457F,
                ),
                call.StorageDiff(
                    key=0x41FB32DAA5E26566D482776D5F8ABFF389DA500038A86A288F5ADB39276AD55,
                    value=0x27FC7A436ABEEE,
                ),
                call.StorageDiff(
                    key=0x5A827682859FAFF64442160965F441FC4344F998E6818244814B04A830749EA,
                    value=0x6380C4AD,
                ),
                call.StorageDiff(
                    key=0x69010E0D78ABB0652354B78994622E24A9E47A8D30D5C4C0C9FB2DF3DBC04CF,
                    value=0xDE2B4696C7152F6,
                ),
                call.StorageDiff(
                    key=0x747A639BB6DBAE6D3A00A22E5E976EB21F6A8C0D87B25A9E77C50759D7066CB,
                    value=0x9737843489E05,
                ),
            ],
        },
        pending_deployed=[],
        pending_nonces={},
        contract_address=3193647238523375300127598243000966435117811015280340811133237455989949224134,
        calldata=[],
        entry_point_selector=227334030968744315992796982100494617316223563777432855541120004521101595501,
        gas_price=0,
    )
```

The state update data above is the minimal subset of the state update for the pending block that still triggers this issue.

The exception we get is:
```
E           starkware.starkware_utils.error_handling.StarkException: (500, {'code': <StarknetErrorCode.TRANSACTION_FAILED: 44>, 'message': 'Error at pc=0:128:\nGot an exception while executing a hint.\nCairo traceback (most recent call last):\nUnknown location (pc=0:3188)\nUnknown location (pc=0:3171)\nUnknown location (pc=0:2163)\nUnknown location (pc=0:992)\n\nError in the called contract (0x70f8a4fcd75190661ca09a7300b7c93fab93971b67ea712c664d7948a8a54c6):\nError at pc=0:128:\nGot an exception while executing a hint.\nCairo traceback (most recent call last):\nUnknown location (pc=0:3409)\nUnknown location (pc=0:3387)\nUnknown location (pc=0:520)\n\nError in the called contract (0x3d39f7248fb2bfb960275746470f7fb470317350ad8656249ec66067559e892):\nError message: SafeUint256: subtraction overflow\nError at pc=0:527:\nAn ASSERT_EQ instruction failed: 0 != 1.\nCairo traceback (most recent call last):\nUnknown location (pc=0:3354)\nUnknown location (pc=0:3330)\nUnknown location (pc=0:908)'})
```

The interesting part from the exception above is:
```Error in the called contract (0x3d39f7248fb2bfb960275746470f7fb470317350ad8656249ec66067559e892):\nError message: SafeUint256: subtraction overflow```

Initially I thought this might maybe be some state in the pending block triggering the overflow. Then I tried to do the following experiment: since we know that the same call works OK for block 11487 (the one having the mint transaction applied) I just took the state update from block 11487 and used that as the pending state update on top of block 11486. This should have caused an equivalent behavior (compared to running the call on 11487) since the resulting state the call is executed on should be the same. Unfortunately the call to the contract above still fails with the complete state update from the next block applied.

This behavior seems to suggest we have some problems when we're applying state updates from the pending block.

--- a/crates/pathfinder/examples/estimate_past_transactions.rs
+++ b/crates/pathfinder/examples/estimate_past_transactions.rs
@@ -243,7 +243,7 @@ async fn estimate(
                 match next_work {
                     Some(Work {transaction, at_block, gas_price, actual_fee, span}) => {
                         let outer = span.clone();
-                        let fut = handle.estimate_fee(transaction, at_block.into(), gas_price, None);
+                        let fut = handle.estimate_fee(transaction, at_block.into(), gas_price, None, None);
                         waiting.push(async move {
                             ReadyResult {
                                 actual_fee,

--- a/crates/pathfinder/src/cairo/ext_py.rs
+++ b/crates/pathfinder/src/cairo/ext_py.rs
@@ -15,7 +15,7 @@
 //! global_state, and after that, calls can be made to it's `block_hash` for which we probably need
 //! to add an alternative way to use a hash directly rather as a root than assume it's a block hash.
 
-use crate::core::CallResultValue;
+use crate::core::{CallResultValue, StarknetBlockTimestamp};
 use crate::rpc::v01::types::{reply::FeeEstimate, request::Call};
 use crate::rpc::v02::types::request::{BroadcastedInvokeTransaction, BroadcastedTransaction};
 use crate::sequencer::reply::StateUpdate;
@@ -52,6 +52,7 @@ impl Handle {
         call: Call,
         at_block: BlockHashNumberOrLatest,
         diffs: Option<Arc<StateUpdate>>,
+        block_timestamp: Option<StarknetBlockTimestamp>,
     ) -> Result<Vec<CallResultValue>, CallFailure> {
         use tracing::field::Empty;
         let (response, rx) = oneshot::channel();
@@ -65,6 +66,7 @@ impl Handle {
                     at_block,
                     chain: self.chain,
                     diffs,
+                    block_timestamp,
                     response,
                 },
                 continued_span,
@@ -87,6 +89,7 @@ impl Handle {
         at_block: BlockHashNumberOrLatest,
         gas_price: GasPriceSource,
         diffs: Option<Arc<StateUpdate>>,
+        block_timestamp: Option<StarknetBlockTimestamp>,
     ) -> Result<FeeEstimate, CallFailure> {
         use tracing::field::Empty;
         let (response, rx) = oneshot::channel();
@@ -157,6 +160,7 @@ impl Handle {
                     gas_price,
                     chain: self.chain,
                     diffs,
+                    block_timestamp,
                     response,
                 },
                 continued_span,
@@ -242,6 +246,7 @@ enum Command {
         at_block: BlockHashNumberOrLatest,
         chain: UsedChain,
         diffs: Option<Arc<StateUpdate>>,
+        block_timestamp: Option<StarknetBlockTimestamp>,
         response: oneshot::Sender<Result<Vec<CallResultValue>, CallFailure>>,
     },
     EstimateFee {
@@ -251,6 +256,7 @@ enum Command {
         gas_price: GasPriceSource,
         chain: UsedChain,
         diffs: Option<Arc<StateUpdate>>,
+        block_timestamp: Option<StarknetBlockTimestamp>,
         response: oneshot::Sender<Result<FeeEstimate, CallFailure>>,
     },
 }
@@ -448,6 +454,7 @@ mod tests {
                                 StarkHash::from_be_slice(&b"some blockhash somewhere"[..]).unwrap(),
                             ).into(),
                             None,
+                            None,
                         ).await.unwrap();
                     }
                 })
@@ -517,6 +524,7 @@ mod tests {
                 crate::core::StarknetBlockNumber::new_or_panic(1).into(),
                 super::GasPriceSource::PastBlock,
                 None,
+                None,
             )
             .await
             .unwrap();
@@ -540,6 +548,7 @@ mod tests {
                 )
                 .into(),
                 super::GasPriceSource::Current(H256::from_low_u64_be(10)),
+                None,
                 None,
             )
             .await
@@ -609,6 +618,7 @@ mod tests {
                 transaction.clone(),
                 crate::core::StarknetBlockNumber::new_or_panic(1).into(),
                 super::GasPriceSource::PastBlock,
+                None,
                 None,
             )
             .await
@@ -683,6 +693,7 @@ mod tests {
                 )
                 .into(),
                 None,
+                None,
             )
             .await
             .unwrap_err();
@@ -748,6 +759,7 @@ mod tests {
                 call.clone(),
                 crate::rpc::v01::types::Tag::Latest.try_into().unwrap(),
                 None,
+                None,
             )
             .await
             .unwrap();
@@ -781,6 +793,7 @@ mod tests {
                 call,
                 crate::rpc::v01::types::Tag::Latest.try_into().unwrap(),
                 Some(update),
+                None,
             )
             .await
             .unwrap();

--- a/crates/pathfinder/src/cairo/ext_py/ser.rs
+++ b/crates/pathfinder/src/cairo/ext_py/ser.rs
@@ -39,6 +39,7 @@ pub(crate) struct CommonProperties<'a> {
     pub pending_updates: ContractUpdatesWrapper<'a>,
     pub pending_deployed: DeployedContractsWrapper<'a>,
     pub pending_nonces: NoncesWrapper<'a>,
+    pub pending_timestamp: u64,
 }
 
 /// Private version of [`crate::core::Chain`] for serialization.

--- a/crates/pathfinder/src/cairo/ext_py/sub_process.rs
+++ b/crates/pathfinder/src/cairo/ext_py/sub_process.rs
@@ -335,6 +335,7 @@ async fn process(
             at_block,
             chain,
             diffs: maybe_diffs,
+            block_timestamp,
             ..
         } => ChildCommand::Call {
             common: CommonProperties {
@@ -343,6 +344,7 @@ async fn process(
                 pending_updates: maybe_diffs.as_ref().map(|x| &**x).into(),
                 pending_deployed: maybe_diffs.as_ref().map(|x| &**x).into(),
                 pending_nonces: maybe_diffs.as_ref().map(|x| &**x).into(),
+                pending_timestamp: block_timestamp.map(|t| t.get()).unwrap_or_default(),
             },
             contract_address: &call.contract_address,
             calldata: &call.calldata,
@@ -354,6 +356,7 @@ async fn process(
             gas_price,
             chain,
             diffs: maybe_diffs,
+            block_timestamp,
             ..
         } => ChildCommand::EstimateFee {
             common: CommonProperties {
@@ -362,6 +365,7 @@ async fn process(
                 pending_updates: maybe_diffs.as_ref().map(|x| &**x).into(),
                 pending_deployed: maybe_diffs.as_ref().map(|x| &**x).into(),
                 pending_nonces: maybe_diffs.as_ref().map(|x| &**x).into(),
+                pending_timestamp: block_timestamp.map(|t| t.get()).unwrap_or_default(),
             },
             gas_price: gas_price.as_price(),
             transaction,

--- a/crates/pathfinder/src/rpc/v01/api.rs
+++ b/crates/pathfinder/src/rpc/v01/api.rs
@@ -895,9 +895,12 @@ impl RpcApi {
             .as_ref()
             .ok_or_else(|| internal_server_error("Unsupported configuration"))?;
 
-        let (when, pending_update) = self.base_block_and_pending_for_call(block_id).await?;
+        let (when, pending_timestamp, pending_update) =
+            self.base_block_and_pending_for_call(block_id).await?;
 
-        Ok(handle.call(request, when, pending_update).await?)
+        Ok(handle
+            .call(request, when, pending_update, pending_timestamp)
+            .await?)
     }
 
     /// Get the latest local block's number.
@@ -1444,10 +1447,17 @@ impl RpcApi {
             GasPriceSource::PastBlock
         };
 
-        let (when, pending_update) = self.base_block_and_pending_for_call(block_id).await?;
+        let (when, pending_timestamp, pending_update) =
+            self.base_block_and_pending_for_call(block_id).await?;
 
         Ok(handle
-            .estimate_fee(transaction, when, gas_price, pending_update)
+            .estimate_fee(
+                transaction,
+                when,
+                gas_price,
+                pending_update,
+                pending_timestamp,
+            )
             .await?)
     }
 
@@ -1461,6 +1471,7 @@ impl RpcApi {
     ) -> Result<
         (
             BlockHashNumberOrLatest,
+            Option<StarknetBlockTimestamp>,
             Option<Arc<sequencer::reply::StateUpdate>>,
         ),
         anyhow::Error,
@@ -1468,7 +1479,7 @@ impl RpcApi {
         use crate::cairo::ext_py::Pending;
 
         match BlockHashNumberOrLatest::try_from(at_block) {
-            Ok(when) => Ok((when, None)),
+            Ok(when) => Ok((when, None, None)),
             Err(Pending) => {
                 // we must have pending_data configured for pending requests, otherwise we fail
                 // fast.
@@ -1476,16 +1487,21 @@ impl RpcApi {
 
                 // call on this particular parent block hash; if it's not found at query time over
                 // at python, it should fall back to latest and **disregard** the pending data.
-                let pending_on_top_of_a_block = pending
-                    .state_update_on_parent_block()
-                    .await
-                    .map(|(parent_block, data)| (parent_block.into(), Some(data)));
+                let pending_on_top_of_a_block = pending.state_update_on_parent_block().await.map(
+                    |(parent_block, timestamp, data)| {
+                        (parent_block.into(), Some(timestamp), Some(data))
+                    },
+                );
 
                 // if there is no pending data available, just execute on whatever latest. this is
                 // the "intent" of the pending functinality other rpc methods should follow as
                 // well, that "pending" is just an emphemeral view of the latest, when it's not
                 // available one is supposed to use latest (for example: testnet).
-                Ok(pending_on_top_of_a_block.unwrap_or((BlockHashNumberOrLatest::Latest, None)))
+                Ok(pending_on_top_of_a_block.unwrap_or((
+                    BlockHashNumberOrLatest::Latest,
+                    None,
+                    None,
+                )))
             }
         }
     }

--- a/crates/pathfinder/src/rpc/v02/method/call.rs
+++ b/crates/pathfinder/src/rpc/v02/method/call.rs
@@ -64,12 +64,17 @@ pub async fn call(
         .as_ref()
         .ok_or_else(|| anyhow::anyhow!("Unsupported configuration"))?;
 
-    let (when, pending_update) =
+    let (when, pending_timestamp, pending_update) =
         super::estimate_fee::base_block_and_pending_for_call(input.block_id, &context.pending_data)
             .await?;
 
     let result = handle
-        .call(input.request.into(), when, pending_update)
+        .call(
+            input.request.into(),
+            when,
+            pending_update,
+            pending_timestamp,
+        )
         .await?;
 
     Ok(result)

--- a/crates/pathfinder/src/rpc/v02/method/estimate_fee.rs
+++ b/crates/pathfinder/src/rpc/v02/method/estimate_fee.rs
@@ -5,7 +5,7 @@ use serde_with::serde_as;
 
 use crate::{
     cairo::ext_py::{BlockHashNumberOrLatest, GasPriceSource},
-    core::BlockId,
+    core::{BlockId, StarknetBlockTimestamp},
     rpc::v02::types::request::BroadcastedTransaction,
     rpc::v02::RpcContext,
     state::PendingData,
@@ -69,11 +69,17 @@ pub async fn estimate_fee(
         GasPriceSource::PastBlock
     };
 
-    let (when, pending_update) =
+    let (when, pending_timestamp, pending_update) =
         base_block_and_pending_for_call(input.block_id, &context.pending_data).await?;
 
     let result = handle
-        .estimate_fee(input.request, when, gas_price, pending_update)
+        .estimate_fee(
+            input.request,
+            when,
+            gas_price,
+            pending_update,
+            pending_timestamp,
+        )
         .await?;
 
     Ok(result.into())
@@ -87,6 +93,7 @@ pub(super) async fn base_block_and_pending_for_call(
 ) -> Result<
     (
         BlockHashNumberOrLatest,
+        Option<StarknetBlockTimestamp>,
         Option<Arc<crate::sequencer::reply::StateUpdate>>,
     ),
     anyhow::Error,
@@ -94,7 +101,7 @@ pub(super) async fn base_block_and_pending_for_call(
     use crate::cairo::ext_py::Pending;
 
     match BlockHashNumberOrLatest::try_from(at_block) {
-        Ok(when) => Ok((when, None)),
+        Ok(when) => Ok((when, None, None)),
         Err(Pending) => {
             // we must have pending_data configured for pending requests, otherwise we fail
             // fast.
@@ -105,13 +112,16 @@ pub(super) async fn base_block_and_pending_for_call(
                     let pending_on_top_of_a_block = pending
                         .state_update_on_parent_block()
                         .await
-                        .map(|(parent_block, data)| (parent_block.into(), Some(data)));
+                        .map(|(parent_block, timestamp, data)| {
+                            (parent_block.into(), Some(timestamp), Some(data))
+                        });
 
                     // if there is no pending data available, just execute on whatever latest.
-                    Ok(
-                        pending_on_top_of_a_block
-                            .unwrap_or((BlockHashNumberOrLatest::Latest, None)),
-                    )
+                    Ok(pending_on_top_of_a_block.unwrap_or((
+                        BlockHashNumberOrLatest::Latest,
+                        None,
+                        None,
+                    )))
                 }
                 None => Err(anyhow::anyhow!(
                     "Pending data not supported in this configuration"

--- a/crates/pathfinder/src/state/sync.rs
+++ b/crates/pathfinder/src/state/sync.rs
@@ -8,7 +8,7 @@ use std::sync::Arc;
 use crate::{
     core::{
         Chain, ClassHash, ContractRoot, GasPrice, GlobalRoot, SequencerAddress, StarknetBlockHash,
-        StarknetBlockNumber,
+        StarknetBlockNumber, StarknetBlockTimestamp,
     },
     ethereum::{log::StateUpdateLog, transport::EthereumTransport},
     rpc::v01::types::reply::{syncing, syncing::NumberedBlock, Syncing as SyncStatus},
@@ -85,11 +85,19 @@ impl PendingData {
 
     pub async fn state_update_on_parent_block(
         &self,
-    ) -> Option<(StarknetBlockHash, Arc<sequencer::reply::StateUpdate>)> {
+    ) -> Option<(
+        StarknetBlockHash,
+        StarknetBlockTimestamp,
+        Arc<sequencer::reply::StateUpdate>,
+    )> {
         let g = self.inner.read().await;
         let inner = g.as_ref()?;
 
-        Some((inner.block.parent_hash, inner.state_update.clone()))
+        Some((
+            inner.block.parent_hash,
+            inner.block.timestamp,
+            inner.state_update.clone(),
+        ))
     }
 }
 

--- a/py/src/pathfinder_worker/call.py
+++ b/py/src/pathfinder_worker/call.py
@@ -1,4 +1,5 @@
 import asyncio
+import dataclasses
 import json
 import os
 import re
@@ -129,6 +130,10 @@ class Command:
     def has_pending_data(self):
         pass
 
+    @abstractmethod
+    def get_pending_timestamp(self) -> int:
+        pass
+
 
 @marshmallow_dataclass.dataclass(frozen=True)
 class Call(Command):
@@ -139,6 +144,7 @@ class Call(Command):
     )
     pending_deployed: List[DeployedContract] = field(metadata=pending_deployed_metadata)
     pending_nonces: Dict[int, int] = field(metadata=pending_nonces_metadata)
+    pending_timestamp: int = field(metadata=fields.timestamp_metadata)
 
     contract_address: int = field(metadata=fields.contract_address_metadata)
     calldata: List[int] = field(metadata=fields.call_data_as_hex_metadata)
@@ -155,6 +161,9 @@ class Call(Command):
             or len(self.pending_nonces) > 0
         )
 
+    def get_pending_timestamp(self) -> int:
+        return self.pending_timestamp
+
 
 @marshmallow_dataclass.dataclass(frozen=True)
 class EstimateFee(Command):
@@ -165,6 +174,7 @@ class EstimateFee(Command):
     )
     pending_deployed: List[DeployedContract] = field(metadata=pending_deployed_metadata)
     pending_nonces: Dict[int, int] = field(metadata=pending_nonces_metadata)
+    pending_timestamp: int = field(metadata=fields.timestamp_metadata)
 
     # zero means to use the gas price from the current block.
     gas_price: int = field(metadata=fields.gas_price_metadata)
@@ -177,6 +187,9 @@ class EstimateFee(Command):
             or len(self.pending_deployed) > 0
             or len(self.pending_nonces) > 0
         )
+
+    def get_pending_timestamp(self) -> int:
+        return self.pending_timestamp
 
 
 class CommandSchema(marshmallow_oneofschema.OneOfSchema):
@@ -369,6 +382,11 @@ def loop_inner(connection, command: Command):
             )
         else:
             raise
+
+    if command.get_pending_timestamp():
+        block_info = dataclasses.replace(
+            block_info, block_timestamp=command.get_pending_timestamp()
+        )
 
     timings["resolve_block"] = time.time() - started_at
     started_at = time.time()

--- a/py/tests/pathfinder_worker/test_call.py
+++ b/py/tests/pathfinder_worker/test_call.py
@@ -38,6 +38,7 @@ def test_command_parsing_estimate_fee():
             }
         ],
         "pending_nonces":{"0x123":"0x1"},
+        "pending_timestamp": 0,
         "transaction":{
             "type":"INVOKE_FUNCTION",
             "version":"0x100000000000000000000000000000000",
@@ -65,6 +66,7 @@ def test_command_parsing_estimate_fee():
             )
         ],
         pending_nonces={0x123: 1},
+        pending_timestamp=0,
         transaction=InvokeFunction(
             version=0x100000000000000000000000000000000,
             contract_address=0x57DDE83C18C0EFE7123C36A52D704CF27D5C38CDF0B1E1EDC3B0DAE3EE4E374,
@@ -90,6 +92,7 @@ def test_command_parsing_call():
         },
         "pending_deployed":[],
         "pending_nonces":{},
+        "pending_timestamp": 0,
         "contract_address":"0x57dde83c18c0efe7123c36a52d704cf27d5c38cdf0b1e1edc3b0dae3ee4e374",
         "calldata":["0x84"],
         "entry_point_selector":"0x26813d396fdb198e9ead934e4f7a592a8b88a059e45ab0eb6ee53494e8d45b0"
@@ -105,6 +108,7 @@ def test_command_parsing_call():
         },
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
         contract_address=0x57DDE83C18C0EFE7123C36A52D704CF27D5C38CDF0B1E1EDC3B0DAE3EE4E374,
         calldata=[0x84],
         entry_point_selector=0x26813D396FDB198E9EAD934E4F7A592A8B88A059E45AB0EB6EE53494E8D45B0,
@@ -340,7 +344,7 @@ def test_success():
     contract_address = hex(populate_test_contract_with_132_on_3(con))
     entry_point = hex(get_selector_from_name("get_value"))
 
-    common_command_data = f'"contract_address": "{contract_address}", "entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}'
+    common_command_data = f'"contract_address": "{contract_address}", "entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}, "pending_timestamp": 0'
 
     output = default_132_on_3_scenario(
         con,
@@ -374,6 +378,7 @@ def test_positive_directly():
         pending_updates={},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     con.execute("BEGIN")
@@ -388,7 +393,7 @@ def test_called_contract_not_found():
     contract_address = populate_test_contract_with_132_on_3(con)
     entry_point = hex(get_selector_from_name("get_value"))
 
-    common_command_data = f'"entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}'
+    common_command_data = f'"entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}, "pending_timestamp": 0'
 
     output = default_132_on_3_scenario(
         con,
@@ -405,7 +410,7 @@ def test_nested_called_contract_not_found():
     contract_address = populate_test_contract_with_132_on_3(con)
     entry_point = hex(get_selector_from_name("call_increase_value"))
 
-    common_command_data = '"gas_price": 0, "chain": "GOERLI", "pending_updates": {}, "pending_deployed": [], "pending_nonces": {}'
+    common_command_data = '"gas_price": 0, "chain": "GOERLI", "pending_updates": {}, "pending_deployed": [], "pending_nonces": {}, "pending_timestamp": 0'
 
     output = default_132_on_3_scenario(
         con,
@@ -424,7 +429,7 @@ def test_invalid_entry_point():
     contract_address = populate_test_contract_with_132_on_3(con)
     entry_point = hex(get_selector_from_name("call_increase_value2"))
 
-    common_command_data = '"gas_price": 0, "chain": "GOERLI", "pending_updates": {}, "pending_deployed": [], "pending_nonces": {}'
+    common_command_data = '"gas_price": 0, "chain": "GOERLI", "pending_updates": {}, "pending_deployed": [], "pending_nonces": {}, "pending_timestamp": 0'
     output = default_132_on_3_scenario(
         con,
         [
@@ -444,7 +449,7 @@ def test_invalid_schema_version():
     contract_address = hex(populate_test_contract_with_132_on_3(con))
     entry_point = hex(get_selector_from_name("get_value"))
 
-    common_command_data = f'"entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}'
+    common_command_data = f'"entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}, "pending_timestamp": 0'
 
     con.execute("pragma user_version = 0")
     con.commit()
@@ -464,7 +469,7 @@ def test_no_such_block():
     contract_address = hex(populate_test_contract_with_132_on_3(con))
     entry_point = hex(get_selector_from_name("get_value"))
 
-    common_command_data = f'"contract_address": "{contract_address}", "entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}'
+    common_command_data = f'"contract_address": "{contract_address}", "entry_point_selector": "{entry_point}", "calldata": ["0x84"], "gas_price": 0, "chain": "GOERLI", "pending_updates": {{}}, "pending_deployed": [], "pending_nonces": {{}}, "pending_timestamp": 0'
 
     con.execute("delete from starknet_blocks")
     con.commit()
@@ -508,6 +513,7 @@ def test_fee_estimate_on_positive_directly():
         pending_updates={},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
         transaction=InvokeFunction(
             version=0x100000000000000000000000000000000,
             contract_address=contract_address,
@@ -551,6 +557,7 @@ def test_fee_estimate_for_declare_transaction_directly():
         pending_updates={},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
         transaction=Declare(
             version=0x100000000000000000000000000000000,
             max_fee=0,
@@ -583,6 +590,7 @@ def test_fee_estimate_on_positive():
         "pending_updates":{{}},
         "pending_deployed":[],
         "pending_nonces":{{}},
+        "pending_timestamp":0,
         "transaction":{{
             "type":"INVOKE_FUNCTION",
             "version":"0x100000000000000000000000000000000",
@@ -669,6 +677,7 @@ def test_call_on_pending_updated():
         pending_updates={contract_address: [call.StorageDiff(key=0x84, value=0x99)]},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     (verb, output, _timings) = loop_inner(con, command)
@@ -698,6 +707,7 @@ def test_call_on_pending_deployed():
             call.DeployedContract(address=contract_address, contract_hash=contract_hash)
         ],
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     (verb, output, _timings) = loop_inner(con, command)
@@ -734,6 +744,7 @@ def test_call_on_pending_deployed_through_existing():
             call.DeployedContract(address=contract_address, contract_hash=contract_hash)
         ],
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     # the call_increase_value doesn't return anything, which is a bit unfortunate.
@@ -779,6 +790,7 @@ def test_call_on_reorgged_pending_block():
                 },
                 pending_deployed=[],
                 pending_nonces={},
+                pending_timestamp=0,
             ),
             [5],
         ),
@@ -796,6 +808,7 @@ def test_call_on_reorgged_pending_block():
                 },
                 pending_deployed=[],
                 pending_nonces={},
+                pending_timestamp=0,
             ),
             [3],
         ),
@@ -815,6 +828,7 @@ def test_call_on_reorgged_pending_block():
                     )
                 ],
                 pending_nonces={},
+                pending_timestamp=0,
             ),
             [5],
         ),
@@ -835,6 +849,7 @@ def test_call_on_reorgged_pending_block():
                     )
                 ],
                 pending_nonces={},
+                pending_timestamp=0,
             ),
             "StarknetErrorCode.UNINITIALIZED_CONTRACT",
         ),
@@ -1033,6 +1048,7 @@ def test_nonce_with_dummy():
         pending_updates={},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
         transaction=base_transaction,
     )
 
@@ -1151,6 +1167,7 @@ def test_failing_mainnet_tx2():
         pending_updates={},
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
         transaction=InvokeFunction(
             version=0,
             contract_address=45915111574649954983606422480229741823594314537836586888051448850027079668,
@@ -1222,6 +1239,7 @@ def test_positive_streamed_on_early_goerli_block_without_deployed():
         },
         pending_deployed=[],
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     without_updates = dataclasses.replace(with_updates, pending_updates={})
@@ -1231,6 +1249,53 @@ def test_positive_streamed_on_early_goerli_block_without_deployed():
 
     (verb, output, _timings) = loop_inner(con, with_updates)
     assert output == [0x22B]
+
+
+@pytest.mark.skip(reason="this requires a mainnet database with block 11486")
+def test_timestamp_dependent_pending_call():
+    con = sqlite3.connect(test_relative_path("../../../mainnet.sqlite"))
+    con.execute("BEGIN")
+
+    # This call fails if we are just using the timestamp from the block specified
+    # by `at_block`. Execution succeeds only with the correct timestamp from the pending
+    # block.
+    c = Call(
+        at_block="0x74cc50f4b8083835682bbe1489a4b1474053e4f6a35c66b8bb25b2e700630f9",
+        chain=call.Chain.MAINNET,
+        pending_updates={
+            0x3D39F7248FB2BFB960275746470F7FB470317350AD8656249EC66067559E892: [
+                call.StorageDiff(
+                    key=0x57E3AC4C831E9CE78DE8142D212350256E7263262456048615D1800F91A021,
+                    value=0xDE21C10E8E2457F,
+                ),
+                call.StorageDiff(
+                    key=0x41FB32DAA5E26566D482776D5F8ABFF389DA500038A86A288F5ADB39276AD55,
+                    value=0x27FC7A436ABEEE,
+                ),
+                call.StorageDiff(
+                    key=0x5A827682859FAFF64442160965F441FC4344F998E6818244814B04A830749EA,
+                    value=0x6380C4AD,
+                ),
+                call.StorageDiff(
+                    key=0x69010E0D78ABB0652354B78994622E24A9E47A8D30D5C4C0C9FB2DF3DBC04CF,
+                    value=0xDE2B4696C7152F6,
+                ),
+                call.StorageDiff(
+                    key=0x747A639BB6DBAE6D3A00A22E5E976EB21F6A8C0D87B25A9E77C50759D7066CB,
+                    value=0x9737843489E05,
+                ),
+            ],
+        },
+        pending_deployed=[],
+        pending_nonces={},
+        pending_timestamp=1669383496,
+        contract_address=3193647238523375300127598243000966435117811015280340811133237455989949224134,
+        calldata=[],
+        entry_point_selector=227334030968744315992796982100494617316223563777432855541120004521101595501,
+        gas_price=0,
+    )
+
+    (verb, output, _timings) = loop_inner(con, c)
 
 
 @pytest.mark.skip(reason="this requires an early goerli database")
@@ -1275,6 +1340,7 @@ def test_positive_streamed_on_early_goerli_block_with_deployed():
         pending_updates=pending_updates,
         pending_deployed=pending_deployed,
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     without_updates = dataclasses.replace(
@@ -1304,6 +1370,7 @@ def test_positive_streamed_on_early_goerli_block_with_deployed():
         pending_updates=pending_updates,
         pending_deployed=pending_deployed,
         pending_nonces={},
+        pending_timestamp=0,
     )
 
     (verb, output, _timings) = loop_inner(con, on_newly_deployed)

